appsrc/dcfile/dcsrdump.man
.TH DCSRDUMP 1 "18 July 2009" "DICOM PS3" "DICOM PS3 - Describe DICOM SR content"
.SH NAME
dcsrdump \- ACR/NEMA DICOM PS3 ... DICOM PS3 - Describe DICOM SR content
.SH SYNOPSIS
.HP 10
.B dcsrdump
[
.B \-v|verbose
]
[
.B \-vv|veryverbose
]
[
.B \-filename
]
[
.B \-identifier
]
.so man1/optin.so
.SH DESCRIPTION
.LP
.B dcsrdump
reads the named dicom structured report (SR) input file and describes the information contained.
.SH OPTIONS
The attribute values, description and verbose output go to standard error.
.PP
The basic input switches are described in dcintro(1). Options specific to this program are:
.TP
.B \-verbose
.RS
Describe the parsing process; does not actually do anything at present.
.RE
.TP
.B \-veryverbose
.RS
Display byte offset from file start and message start, in hexadecimal, and dump contents of the DICOM dataset while reading.
.RE
.TP
.B \-filename
.RS
Show the name of the file supplied in the arguments; a hyphen will be reported if no filename was supplied.
.RE
.TP
.B \-identifier
.RS
Preceed each content item with the dot delimited numeric identifier of the content item (as used in references).
.RE
.SH ENVIRONMENT
.LP
\ 
.SH EXAMPLES
.LP
.RE
\ 
.RE
% dcsrdump -identifier basictextsr.dcm 
.RE
1: : CONTAINER: (18748-4,LN,"Diagnostic Imaging Report")  [SEPARATE]
.RE
	>1.1: CONTAINS: CONTAINER: (121070,DCM,"Findings")  [SEPARATE]
.RE
		>>1.1.1: CONTAINS: TEXT: (121071,DCM,"Finding")  = "stuff"
.RE
.SH FILES
.LP
\ 
.SH SEE ALSO
.BR dcintro(1)
.SH AUTHOR
Copyright (C) 1993-2010. David A. Clunie DBA PixelMed Publishing. All rights reserved.
.SH BUGS


CHANGES
111219:	libsrc/src/dctool/attrmxrd.cc: check for HighBit being more than one less than BitsStored and correct BitsStored accordingly (otherwise will create zero pixel values on copy) (000408)

111213:	appsrc/dcfile/dcuncat.cc: add option to replace instance number starting with a specific number to allow scripts to extend a series with successive invocations

111212:	libsrc/standard/condn.tpl/module,strval/base.tpl,dx.tpl: improve animal versus human conditions to check SRT homo sapiens, and make view position and body part examined human versus animal specific

111212:	libsrc/standard/elmdict/other.tpl: add more Sound private elements

111117:	libsrc/standard/strval/base.tpl: add CSD for 99NCIAIM

111109:	appsrc/dcfile/dcuncat.cc: add options to replace instance number, frame increment pointer and frame increment pointer vectors, as well as add number of frames during unenhance based on SOP class

111106:	libsrc/standard/elmdict/other.tpl: add Vepro private elements

111031:	libsrc/standard/elmdict/other.tpl: add CTP private elements

111030:	libsrc/standard/elmdict/other.tpl: add Segami private elements

111029:	libsrc/standard/elmdict/dicom3.tpl: add (restore from trial) Observation UID (per CP 1147)

111028:	libsrc/standard/elmdict/other.tpl: add NeuroQuant private elements

111024:	libsrc/standard/elmdict/other.tpl: update fuji private attributes with descriptions

110918:	libsrc/standard/elmdict/diconde.tpl: update with new or modified data elements with cleanup from Patrick

110918:	libsrc/standard/elmdict/siemens.tpl: update some private elements based on Artis conformance statement

110916:	libsrc/standard/elmdict/gems.tpl: resolve inconsistencies in GE PET conformance statement VRs SS/SL/FL for private elements by choosing those most commonly seen in the field (still some inconsistencies between models)

110904:	libsrc/standard/elmdict/camtron.tpl: add private data elements encountered in DICOMDIR, as well as QCA private data elements from GEMS/Camtronics CRS conformance statement

110901:	libsrc/standard/sopcl.tpl,elmdict/dicom3.tpl: Sup 152 FT ophthalmic thickness map (dictionary and SOP class only; not IOD yet)

110822:	libsrc/standard/elmdict/dicom3.tpl: update data dictionary (only) for CPs for FT Aug 2011 WG 6 meeting

110714:	libsrc/standard/elmdict/dicom3.tpl: remove 'unified' from names of various attributes (per Sup 96 FT)

110716:	libsrc/standard/strval/base.tpl: add 99PMP CSD

110714:	libsrc/standard/elmdict/dicom3.tpl: exposure index etc are DS not FD (CP 1024)

110709:	libsrc/standard/elmdict/dicom3.tpl: change (0018,9430) from US to FL per Sup 139

110612:	libsrc/standard/elmdict/diconde.tpl: update for DICONDE eddy current standard

110612:	libsrc/standard/elmdict/dicondep.tpl: correct case of DICONDE DX (Dx) private owner for E2699

110609:	libsrc/standard/elmdict/dicom3.tpl: add Sup 151 FT data elements

110608:	libsrc/standard/elmdict/elscint.tpl: update Elscint private elements

110531:	libsrc/standard/elmdict/other.tpl: update Hitachi CT private elements

110531:	libsrc/standard/elmdict/siemens.tpl: update mammo private elements

110529:	libsrc/standard/condn.tpl,iodcomp/dx.tpl,module/base.tpl,dx.tpl,xaxrf.tpl,strval/dx.tpl,xaxrf.tpl: add breast tomosynthesis IOD, refactoring 3D XRay accordingly

110529:	libsrc/standard/elmdict/other.tpl: more private data elements for Hologic 

110528:	appsrc/dcfile/dciodvfy.cc: do not complain about SCOORD without child IMAGE when SELECTED FROM is by reference (000406)

110527:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: check that references to instances in content tree are in top level evidence sequences in SR (000405)

110526:	libsrc/standard/elmdict/gems.tpl: add names for CT iterative reconstruction private data elements

110522:	libsrc/src/dctool/attrothr.cc: do not throw assertion if computed length is equal (which may occur for single bit images)

110522:	libsrc/standard/iodcomp/base.tpl,libsrc/standard/condn.tpl: permit FoR Module in Segementation (and do CP to standard to allow may be present otherwise)

110522:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: detect SCOORD without child IMAGE (000403)

110522:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: detect quotes in code meaning of measurement units (000404)

110522:	libsrc/src/dctool/attrtypv.cc,locale/mesgtext.tpl: permit PN with multiple carets in multiple component groups (000402)

110521:	libsrc/standard/elmdict/acuson.tpl,agfa.tpl,isg.tpl,other.tpl,siemens.tpl,strval/base.tpl: more private elements and coding scheme designators

110517:	libsrc/standard/module/base.tpl: correct type (1->2) and number of items (1->0-n) in SourceImageSequence in DerivationImageMacro

110517:	libsrc/standard/elmdict/siemens.tpl: yet more syngo private elements

110516:	libsrc/standard/cond.tpl: no longer require version for UCUM codes (CP 1060)

110516:	libsrc/standard/strval/base.tpl: add 99_MI private coding scheme designator

110516:	libsrc/include/dctool/attr.h,attrseq.h,generic/listsimp.h,src/dctool/attrseq.cc,attrvrfy.cc,standard/condn.tpl,module/softcopy.tpl,sr.tpl,support/module.awk:
		though VM for sequences number of items was being specified in modules, it was not being applied;
		apply it (000256), as well as make it work (since number of items was not being computed correctly
		in SequenceAttribute::verifyVM()); also add support for conditional check of VM, for example for
		GraphicData depending on Grahic Type (000401)

110516:	libsrc/standard/elmdict/toshiba.tpl: update CT private elements from conformance statement for V4.70 or later version

110514:	libsrc/standard/elmdict/siemens.tpl: add more syngo private elements, and update guessed VRs of existing ones with observed explicit VRs

110511:	libsrc/standard/elmdict/gems.tpl,toshiba.tpl: add more encountered explicit VR private elements

110511:	libsrc/standard/elmdict/other.tpl: update/add McKesson private elements from conformance statement


appsrc/dcfile/dcsrdump.cc
#include "attrmxls.h"
#include "mesgtext.h"
#include "dcopt.h"
#include "elmconst.h"
#include "attrval.h"

static inline void
indent(TextOutputStream &log,unsigned depth,bool showtabs,bool showgt) {
	unsigned d=depth;
	if (showtabs) while (d--) log << "\t";
	d=depth;
	if (showgt) while (d--) log << ">";
}

static void
logStringValueOfAttribute(Attribute *a,TextOutputStream &log) {
	if (a) {
		char *v = AttributeValue(a);
		if (v) {
			log << v;
			delete[] v;
		}
	}
}

static void
processCodeSequenceItem(AttributeList *list,TextOutputStream &log) {
	log << "(";
	logStringValueOfAttribute((*list)[TagFromName(CodeValue)],log);
	log << ",";
	logStringValueOfAttribute((*list)[TagFromName(CodingSchemeDesignator)],log);
	log << "," << "\"" ;
	logStringValueOfAttribute((*list)[TagFromName(CodeMeaning)],log);
	log << "\")";
}

static void
processCodeSequence(Attribute *aCodeSequence,const char *label,TextOutputStream &log) {
	log << label;

	if (aCodeSequence) {
		int nItems = 0;
		AttributeList ** vItems = 0;
		nItems=aCodeSequence->getLists(&vItems);
		Assert(nItems==0 || vItems);

		int i;
		AttributeList **iptr;

		for (i=0,iptr=vItems; i<nItems; ++i,++iptr) {
			AttributeList *ptr = *iptr;
			Assert(ptr);
			if (i > 0) log << ", ";
			processCodeSequenceItem(ptr,log);
		}
	}
}

static void
processMeasuredValueSequence(Attribute *aMeasuredValueSequence,const char *label,TextOutputStream &log) {
	log << label;

	if (aMeasuredValueSequence) {
		int nItems = 0;
		AttributeList ** vItems = 0;
		nItems=aMeasuredValueSequence->getLists(&vItems);
		Assert(nItems==0 || vItems);

		int i;
		AttributeList **iptr;

		for (i=0,iptr=vItems; i<nItems; ++i,++iptr) {
			AttributeList *ptr = *iptr;
			Assert(ptr);
			if (i > 0) log << ", ";

			logStringValueOfAttribute((*ptr)[TagFromName(NumericValue)],log);

			Attribute *aMeasurementUnitsCodeSequence=(*ptr)[TagFromName(MeasurementUnitsCodeSequence)];
			if (aMeasurementUnitsCodeSequence) {
				log << " ";
				processCodeSequence(aMeasurementUnitsCodeSequence,"",log);
			}
		}
	}
}

static void
processReferencedSOPSequence(Attribute *aReferencedSOPSequence,const char *label,TextOutputStream &log) {
	log << label;

	if (aReferencedSOPSequence) {
		int nItems = 0;
		AttributeList ** vItems = 0;
		nItems=aReferencedSOPSequence->getLists(&vItems);
		Assert(nItems==0 || vItems);

		int i;
		AttributeList **iptr;

		for (i=0,iptr=vItems; i<nItems; ++i,++iptr) {
			AttributeList *ptr = *iptr;
			Assert(ptr);
			if (i > 0) log << ", ";

			log << "(";
			logStringValueOfAttribute((*ptr)[TagFromName(ReferencedSOPClassUID)],log);
			log << ",";
			logStringValueOfAttribute((*ptr)[TagFromName(ReferencedSOPInstanceUID)],log);
			log << ")";
			
			// the following items are specific to the usage in IMAGE content items (and hence should do nothing in others) ...
			
			Attribute *aReferencedFrameNumber=(*ptr)[TagFromName(ReferencedFrameNumber)];
			if (aReferencedFrameNumber) {
				log << " [Frame ";
				logStringValueOfAttribute(aReferencedFrameNumber,log);
				log << "]";
			}
			
			Attribute *aReferencedSegmentNumber=(*ptr)[TagFromName(ReferencedSegmentNumber)];
			if (aReferencedSegmentNumber) {
				log << " [Segment ";
				logStringValueOfAttribute(aReferencedSegmentNumber,log);
				log << "]";
			}

			Attribute *aPresentationStateReferencedSOPSequence=(*ptr)[TagFromName(ReferencedSOPSequence)];
			if (aPresentationStateReferencedSOPSequence) {
				int nItems = 0;
				AttributeList ** vItems = 0;
				nItems=aPresentationStateReferencedSOPSequence->getLists(&vItems);
				Assert(nItems==0 || vItems);
				Assert(nItems==0 || vItems);

				int i;
				AttributeList **iptr;

				for (i=0,iptr=vItems; i<nItems; ++i,++iptr) {
					AttributeList *ptr = *iptr;
					Assert(ptr);
					if (i > 0) log << ",";

					log << " (PS ";
					logStringValueOfAttribute((*ptr)[TagFromName(ReferencedSOPClassUID)],log);
					log << ",";
					logStringValueOfAttribute((*ptr)[TagFromName(ReferencedSOPInstanceUID)],log);
					log << ")";
				}
			}

			Attribute *aReferencedRealWorldValueMappingInstanceSequence=(*ptr)[TagFromName(ReferencedRealWorldValueMappingInstanceSequence)];
			if (aReferencedRealWorldValueMappingInstanceSequence) {
				int nItems = 0;
				AttributeList ** vItems = 0;
				nItems=aReferencedRealWorldValueMappingInstanceSequence->getLists(&vItems);
				Assert(nItems==0 || vItems);
				Assert(nItems==0 || vItems);

				int i;
				AttributeList **iptr;

				for (i=0,iptr=vItems; i<nItems; ++i,++iptr) {
					AttributeList *ptr = *iptr;
					Assert(ptr);
					if (i > 0) log << ",";

					log << " (RWV ";
					logStringValueOfAttribute((*ptr)[TagFromName(ReferencedSOPClassUID)],log);
					log << ",";
					logStringValueOfAttribute((*ptr)[TagFromName(ReferencedSOPInstanceUID)],log);
					log << ")";
				}
			}
			
		}
	}
}

static void
processScoordContentItem(AttributeList *list,TextOutputStream &log) {
	logStringValueOfAttribute((*list)[TagFromName(GraphicType)],log);
	log << " {";
	//log.unsetf(ios_base::fixed,ios_base::scientific);	// :) should really explicitly establish the default floating point notation, in which the precision field specifies the maximum number of meaningful digits to display in total counting both those before and those after the decimal point
	log << setprecision (15);							// this choice of precision is based on the fact that a DS VR is 16 characters, including the decimal point; this is excessive in that the FL VR (32 bit IEEE float) significand has a precision of 24 binary bits (about 7 decimal digits)
	Attribute *aGraphicData=(*list)[TagFromName(GraphicData)];
	int nGraphicData = aGraphicData ? aGraphicData->getVM() : 0;
	int i;
	for (i=0; i<nGraphicData; ++ i) {
		if (i > 0) log << ",";
		float v;
		if (aGraphicData->getValue(i,v)) log << v;
	}
	log << "}";
}

static void
processValueOfContentItem(AttributeList *list,TextOutputStream &log) {
	Attribute *aValueType=(*list)[TagFromName(ValueType)];
	char *vValueType=AttributeValue(aValueType);
	if (vValueType) {
		if      (strcmp(vValueType,"NUM") == 0) {
			log << " = ";
			processMeasuredValueSequence((*list)[TagFromName(MeasuredValueSequence)],"",log);
			Attribute *aNumericValueQualifierCodeSequence=(*list)[TagFromName(NumericValueQualifierCodeSequence)];
			if (aNumericValueQualifierCodeSequence) {
				log << " ";
				processCodeSequence(aNumericValueQualifierCodeSequence,"",log);
			}
		}
		else if (strcmp(vValueType,"TEXT") == 0) {
			log << " = \"";
			logStringValueOfAttribute((*list)[TagFromName(TextValue)],log);
			log << "\"";
		}
		else if (strcmp(vValueType,"CODE") == 0) {
			log << " = ";
			processCodeSequence((*list)[TagFromName(ConceptCodeSequence)],"",log);
		}
		else if (strcmp(vValueType,"DATE") == 0) {
			log << " = \"";
			logStringValueOfAttribute((*list)[TagFromName(Date)],log);
			log << "\"";
		}
		else if (strcmp(vValueType,"TIME") == 0) {
			log << " = \"";
			logStringValueOfAttribute((*list)[TagFromName(Time)],log);
			log << "\"";
		}
		else if (strcmp(vValueType,"DATETIME") == 0) {
			log << " = \"";
			logStringValueOfAttribute((*list)[TagFromName(DateTime)],log);
			log << "\"";
		}
		else if (strcmp(vValueType,"PNAME") == 0) {
			log << " = \"";
			logStringValueOfAttribute((*list)[TagFromName(PersonName)],log);
			log << "\"";
		}
		else if (strcmp(vValueType,"UIDREF") == 0) {
			log << " = \"";
			logStringValueOfAttribute((*list)[TagFromName(UID)],log);
			log << "\"";
		}
		else if (strcmp(vValueType,"IMAGE") == 0) {
			log << " = ";
			processReferencedSOPSequence((*list)[TagFromName(ReferencedSOPSequence)],"",log);
		}
		else if (strcmp(vValueType,"WAVEFORM") == 0) {
			processReferencedSOPSequence((*list)[TagFromName(ReferencedSOPSequence)],"",log);
		}
		else if (strcmp(vValueType,"SCOORD") == 0) {
			log << " = ";
			processScoordContentItem(list,log);
		}
		else if (strcmp(vValueType,"TCOORD") == 0) {
		}
		else if (strcmp(vValueType,"COMPOSITE") == 0) {
			processReferencedSOPSequence((*list)[TagFromName(ReferencedSOPSequence)],"",log);
		}
		else if (strcmp(vValueType,"CONTAINER") == 0) {
			log << " [";
			logStringValueOfAttribute((*list)[TagFromName(ContinuityOfContent)],log);
			log << "]";
		}
		else {
			log << " **** Unrecognized Value Type ****";
		}

		delete[] vValueType;
	}
}


static bool
processContentItem(AttributeList *list,TextOutputStream &log,bool verbose,bool veryverbose,unsigned depth,const char *identifierstring)   {

	bool success=true;

	Attribute *        aRelationshipType=(*list)[TagFromName(RelationshipType)];
	Attribute *         aContentSequence=(*list)[TagFromName(ContentSequence)];
	Attribute * aConceptNameCodeSequence=(*list)[TagFromName(ConceptNameCodeSequence)];
	Attribute *               aValueType=(*list)[TagFromName(ValueType)];

	indent(log,depth,true,true);
	if (identifierstring) {
		log << identifierstring << ": ";
	}
	
	logStringValueOfAttribute(aRelationshipType,log);
	log << ": ";

	logStringValueOfAttribute(aValueType,log);
	log << ": ";

	if (aConceptNameCodeSequence) {
		processCodeSequence(aConceptNameCodeSequence,"",log);
		log << " ";
	}

	processValueOfContentItem(list,log);

	log << endl;

	if (aContentSequence) {
		int nContentItems = 0;
		AttributeList ** vContentItems = 0;
		nContentItems=aContentSequence->getLists(&vContentItems);
		Assert(nContentItems==0 || vContentItems);

		int i;
		AttributeList **iptr;

		for (i=0,iptr=vContentItems; i<nContentItems; ++i,++iptr) {
			AttributeList *ptr = *iptr;
			Assert(ptr);
			char *childidentifierstring = NULL;
			if (identifierstring) {
				ostrstream childidentifierstream;
				childidentifierstream << identifierstring << "." << (i+1) << ends;
				childidentifierstring = childidentifierstream.str();
			}
			success&=processContentItem(ptr,log,verbose,veryverbose,depth+1,childidentifierstring);
			if (childidentifierstring) {
				delete [] childidentifierstring;
			}
		}
	}
	return success;
}


static bool
parseStructuredReport(ManagedAttributeList &list,TextOutputStream &log,bool verbose,bool veryverbose,bool showidentifier)
{
	bool result = processContentItem(&list,log,verbose,veryverbose,0,showidentifier ? "1" : NULL);
	return result;
}

int
main(int argc, char *argv[])
{
	GetNamedOptions 	options(argc,argv);
	DicomInputOptions 	dicom_input_options(options);

	bool veryverbose=options.get("veryverbose") || options.get("vv");
	bool verbose=options.get("verbose") || options.get("v") || veryverbose;
	bool showfilename=options.get("filename");
	bool showidentifier=options.get("identifier");

	dicom_input_options.done();
	options.done();

	DicomInputOpenerFromOptions input_opener(
		options,dicom_input_options.filename,cin);

	cerr << dicom_input_options.errors();
	cerr << options.errors();
	cerr << input_opener.errors();

	if (!dicom_input_options.good()
	 || !options.good()
	 || !input_opener.good()
	 || !options) {
		cerr 	<< MMsgDC(Usage) << ": " << options.command()
			<< dicom_input_options.usage()
			<< " [-v|-verbose|-vv|-veryverbose]"
			<< " [-filename]"
			<< " [-identifier]"
			<< " [" << MMsgDC(InputFile) << "]"
			<< " <" << MMsgDC(InputFile)
			<< endl;
		exit(1);
	}

	DicomInputStream din(*(istream *)input_opener,
		dicom_input_options.transfersyntaxuid,
		dicom_input_options.usemetaheader);

	ManagedAttributeList list;

	bool success=true;
	TextOutputStream log(cerr);
	
	if (showfilename) {
		const char *filenameused = input_opener.getFilename();
		cerr << "Filename: \"" << (filenameused && strlen(filenameused) > 0 ? filenameused : "-") << "\"" << endl;
	}

	if (veryverbose) log << "******** While reading ... ********" << endl; 
	list.read(din,&log,veryverbose,0xffffffff,true,dicom_input_options.uselengthtoend,dicom_input_options.ignoreoutofordertags,dicom_input_options.useUSVRForLUTDataIfNotExplicit);

	const char *errors=list.errors();
	if (errors) log << errors << flush;
	if (!list.good()) {
		log << EMsgDC(DatasetReadFailed) << endl;
		success=false;
	}

	//if (veryverbose) log << "******** As read ... ********" << endl; 
	//log << list;
	//list.write(log,veryverbose);

	success=success&&parseStructuredReport(list,log,verbose,veryverbose,showidentifier);

	return success ? 0 : 1;
}




